storage/isolation_manager.rs

```diff
@@ -360,61 +344,70 @@ impl Error for IsolationError {}
         let windows = VecDeque::new();
         Timeline {
-            next_relative_index_and_windows: RwLock::new((next_relative_index, windows)),
-            watermark: AtomicI64::new(next_relative_index - 1),
+            next_sequence_and_windows: RwLock::new((next_sequence_number, windows)),
+            watermark: AtomicU64::new(next_sequence_number.number() - 1),
         }
     }
 
     fn may_free_windows(&self) {
         let watermark = self.watermark();
         let can_free_some: bool = {
-            let (next_relative_index, windows) = &*self.next_relative_index_and_windows.read().unwrap_or_log();
-            let start_of_second_window = *next_relative_index - ((windows.len() - 1) * TIMELINE_WINDOW_SIZE) as i64;
+            let (next_sequence_number, windows) = &*self.next_sequence_and_windows.read().unwrap_or_log();
+            let start_of_second_window: SequenceNumber = Self::start_of_nth_window(windows, *next_sequence_number, 1);
             match windows.front() {
                 None => false,
                 Some(front) => front.get_readers() == 0 && watermark >= start_of_second_window,
             }
         };
         if can_free_some {
-            let (next_relative_index, windows) = &mut *self.next_relative_index_and_windows.write().unwrap_or_log();
-            let mut end_of_first_remaining_window =
-                *next_relative_index - ((windows.len() - 1) * TIMELINE_WINDOW_SIZE) as i64;
-            while watermark >= end_of_first_remaining_window
+            let (next_sequence_number, windows) = &mut *self.next_sequence_and_windows.write().unwrap_or_log();
+            let mut start_of_second_remaining_window: SequenceNumber =
+                Self::start_of_nth_window(windows, *next_sequence_number, 1);
+            while watermark >= start_of_second_remaining_window
                 && windows.front().is_some()
                 && windows.front().unwrap().get_readers() == 0
             {
                 windows.pop_front();
-                end_of_first_remaining_window += TIMELINE_WINDOW_SIZE as i64;
+                start_of_second_remaining_window += TIMELINE_WINDOW_SIZE;
             }
         }
     }
 
-    fn may_increment_watermark(&self, relative_index: i64) {
-        let mut watermark = self.watermark.load(Ordering::SeqCst);
-        if watermark != relative_index - 1 {
+    fn may_increment_watermark(&self, sequence_number: SequenceNumber) {
+        if self.watermark() != sequence_number - 1 {
             return ();
         }
-        let (mut window, mut candidate_slot_index) = self.try_get_window(relative_index).unwrap();
-        let end_of_window: i64 = relative_index + (TIMELINE_WINDOW_SIZE - candidate_slot_index) as i64;
-        let mut candidate_watermark = relative_index;
+        let (mut window, mut candidate_slot_index) = self.try_get_window(sequence_number).unwrap();
+        let end_of_window: SequenceNumber = if candidate_slot_index == 0 {
```

Comment:
Let's only trigger freeing windows on removal of a reader and re-check:
get watermark
loop {
  get first window
  check readers is 0
  check window doesn't contain watermark
  if all 3: delete window
}

Reply:
Can we keep it this way? I like this because the guarantee is: Unused windows are freed.
Not checking on incrementing the watermark also prevents free-as-you-go loading of a committed history.
So we are actually trading off robustness for simplicity.

Reply:
ok we can keep this but we should be able to get rid of `resolve` sequence number stuff - this was the key point of the next comment and it hasn't been implemented?